cores/esp8266/core_esp8266_main.cpp

```diff
@@ -400,79 +400,230 @@ extern "C" void __disableWiFiAtBootTime (void)
+            chip_sz_str = PSTR("Flash size mismatch, check that the build setting matches the device.\n");
+            continue;
+        }
+        #endif
+
+        // All the examples I find, show the partition table in the global address space.
+        static partition_item_t at_partition_table[] =
+        {
+            { SYSTEM_PARTITION_PHY_DATA,         phy_data,         0x1000 }, // type 5
+            { SYSTEM_PARTITION_RF_CAL,           rf_cal,           0x1000 }, // type 4
+            { SYSTEM_PARTITION_SYSTEM_PARAMETER, system_parameter, 0x3000 }, // type 6
+        };
+        _at_partition_table = at_partition_table;
+        _at_partition_table_sz = sizeof(at_partition_table) / sizeof(at_partition_table[0]);
+        // SDK 3.0's `system_partition_table_regist` is FOTA-centric. It will report
+        // on BOOT, OTA1, and OTA2 being missing. We are Non-FOTA. I don't see
+        // anything we can do about this. Other than maybe turning off os_print.
+        if (!system_partition_table_regist(at_partition_table, _at_partition_table_sz, system_get_flash_size_map())) {
+            table_regist_str = PSTR("System partition table registration failed!\n");
+            continue;
+        }
+    } while(false);
+
+    if (chip_sz_str || flash_map_str || table_regist_str) {
+        // user_pre_init() is called very early in the SDK startup. When called,
+        // the PLL CPU clock calibration hasn't not run. Since we are failing, the
+        // calibration will never complete. And the process will repeat over and
+        // over. The effective data rate will always be 74880 bps. If we had a
+        // successful boot, the effective data rate would be 115200 on a restart
+        // or HWDT. This hack relies on the CPU clock calibration never having
+        // completed. This assumes we are starting from a hard reset.
+
+        // A possible exception would be a soft reset after flashing. In which
+        // case the message will not be readable until after a hard reset or
+        // power cycle.
+
+        // After flashing, the Arduino Serial Monitor needs a moment to
+        // reconnect. This also allows time for the FIFO to clear and the host
+        // serial port to clear any framing errors.
+        ets_delay_us(200u * 1000u); // For an uncalibrated CPU Clock, this is close enough.
+
+        #if !defined(F_CRYSTAL)
+        #define F_CRYSTAL 26000000
+        #endif
+        // For print messages to be readable, the UART clock rate is based on the
+        // precalibration rate.
+        if (F_CRYSTAL != 40000000) {
+            uart_div_modify(0, F_CRYSTAL * 2 / 115200);
+            ets_delay_us(150);
+        }
+        do {
+            ETS_PRINTF("\n\n");
+            // Because SDK v3.0.x always has a non-32-bit wide exception handler
+            // installed, we can use PROGMEM strings with Boot ROM print functions.
+#if defined(DEBUG_ESP_CORE) | defined(DEBUG_ESP_PORT) // DEBUG_ESP_CORE => verbose
```

Comment:
```suggestion
#if defined(DEBUG_ESP_CORE) || defined(DEBUG_ESP_PORT) // DEBUG_ESP_CORE => verbose
```
btw whats the reason for ESP PORT dep? another worry is this way we depend on it being properly constructed, which may not be the case if called before c++ ctors init

Reply:
I use `defined(DEBUG_ESP_PORT)` strictly as a debug build flag. It's value is never used for selecting the serial device. All printing is handled through Boot ROM `ets_uart_printf`.

Reply:
I'll correct all the `&` vs `&&` a little bit of language confusion. I took one wrong instance and multiplied it